vm/src/org/mozartoz/truffle/nodes/call/CallMethodNode.java

```diff
@@ -43,89 +46,130 @@ public static CallMethodNode create() {
+		@TruffleBoundary
+		protected static OzProc getMethod(OzObject self, Object name) {
+			DynamicObject methods = (DynamicObject) self.getClazz().get(ooMeth);
+			return (OzProc) methods.get(name);
 		}
 
-		Object[] arguments = new Object[] { self, message };
-		return callNode.call(frame, method.callTarget, OzArguments.pack(method.declarationFrame, arguments));
 	}
 
-	static final OzUniqueName ooMeth = OzUniqueName.get("ooMeth");
+	public static class Lookup {
+		public final OzProc proc;
+		public final boolean embed;
 
-	@TruffleBoundary
-	protected OzProc getMethod(OzObject self, Object name) {
-		DynamicObject methods = (DynamicObject) self.getClazz().get(ooMeth);
-		return (OzProc) methods.get(name);
-	}
+		private Lookup(OzProc proc, boolean embed) {
+			this.proc = proc;
+			this.embed = embed;
+		}
 
-	protected Object labelFromArgs(Object[] args) {
-		return labelNode.executeLabel(derefNode.executeDeref(args[0]));
 	}
 
-	protected OzProc methodFromArgs(OzObject self, Object[] args) {
-		return getMethod(self, labelFromArgs(args));
-	}
+	@ImportStatic(CallMethodNode.class)
+	@NodeChildren({ @NodeChild("self"), @NodeChild("lookup"), @NodeChild("args") })
+	public abstract static class MethodDispatchNode extends OzNode {
+
+		public static MethodDispatchNode create() {
+			return MethodDispatchNodeGen.create(null, null, null);
+		}
+
+		public abstract Object executeDispatch(VirtualFrame frame, OzObject self, Lookup lookup, Object[] args);
+
+		@Specialization(guards = {
+				"lookup.proc == cachedLookup.proc",
+				"!lookup.embed",
+				"!cachedLookup.embed"
+		}, limit = "methodCache(3)")
+		protected Object dispatchCached(VirtualFrame frame, OzObject self, Lookup lookup, Object[] args,
+				@Cached("lookup") Lookup cachedLookup,
+				@Cached("createDirectCallNode(cachedLookup.proc.callTarget)") DirectCallNode callNode) {
+			Object[] arguments = new Object[] { self, args[0] };
+			return callNode.call(frame, OzArguments.pack(cachedLookup.proc.declarationFrame, arguments));
+		}
+
+		@Specialization(guards = {
+				"lookup.proc == cachedLookup.proc",
+				"lookup.embed",
+				"cachedLookup.embed"
+		}, limit = "methodCache(3)")
+		protected Object dispatchCachedOtherwise(VirtualFrame frame, OzObject self, Lookup lookup, Object[] args,
+				@Cached("lookup") Lookup cachedLookup,
+				@Cached("createDirectCallNode(cachedLookup.proc.callTarget)") DirectCallNode callNode) {
+			Object[] arguments = new Object[] { self, OTHERWISE_MESSAGE_FACTORY.newRecord(args[0]) };
+			return callNode.call(frame, OzArguments.pack(cachedLookup.proc.declarationFrame, arguments));
+		}
+
+		BranchProfile otherwiseProfile = BranchProfile.create();
+
+		@Specialization(contains = { "dispatchCached", "dispatchCachedOtherwise" })
+		protected Object dispatchGeneric(VirtualFrame frame, OzObject self, Lookup lookup, Object[] args,
+				@Cached("create()") IndirectCallNode callNode) {
+			Object message = args[0];
+			if (lookup.embed) {
```

Comment:
This should be a ConditionProfile so you can optimize for the case where it's only `otherwise` messages.